Add mortar mechanical contact to ContactAction.

shaneongit opened the pull request:
Closes #13676

Uses the existing Contact action and adds mortar, as requested in the issue.

Things to fix:
- [x] Currently crashes (fixed unsigned vs enum error)
- [x] Add tangential, which should be trivial once normal contact is working.
- [x] Might want to include a postprocessor that dumps the lagrange variable because the other types of contact do that. EDIT: contact pressure aux vars are always added already.
- [ ] Figure out where to put documentation.
- [x] Use LowerDBlockFromSidesetGenerator.
- [ ] Make sure it works with axisymmetric
- [ ] Add extra_vector_tags when #14107 is done.

shaneongit committed:
Add mortar mechanical contact to ContactAction.

Closes #13676

shaneongit commented:
@lindsayad Should the action determine the default order of the lagrange multiplier? If so, should the default be the same as the primal variable or codimension one?

tophmatthews commented:
make sure it works in axisymmetric. Also maybe there is a good place to insert extra_vector_tags for reference residual?

lindsayad commented:
> Should the action determine the default order of the lagrange multiplier? If so, should the default be the same as the primal variable or codimension one?

Now this depends on whether we are doing normal or tangential contact. I've generally found that normal contact works best with a node to face discretization; with that discretization the normal LM variable should live on every node on the slave side so it should have order equal to the primal variable.

For tangential contact, a mortar discretization has worked best. Optimal finite element convergence rates are observed when the tangential LM variable is one order less than the primal variable, so first-order Lagrange if the primal is second-order Lagrange, or constant monomial if the primal variable is first order Lagrange.

shaneongit commented:
@tophmatthews Do you have (or can you point me to) examples of using extra_tag_vectors and/or axisymmetric cases with mortar?

dschwen commented on modules/contact/src/ContactAction.C:
```diff
@@ -231,13 +321,32 @@ ContactAction::getSmoothingEnum()
 }
 
+MooseEnum
+ContactAction::getOrderEnum()
```
why not just call `AddVariableAction::getNonlinearVariableOrders()` directly? That's what we do in all the other actions.

shaneongit replied:
`AddVariableAction::getNonlinearVariableOrders()` sets a default order/family which will be incorrect for tangential mortar constraints. Rather than have a default, I thought it might be better to set the default at runtime based on normal vs tangential and then let the user override it.

dschwen replied:
I don't understand. This method literally just passes the return value `AddVariableAction::getNonlinearVariableOrders();` through.

shaneongit replied:
Sorry, I changed the history of the commit. At the moment, yes, it's just a pass-through. Let me get the variable order working first; then it should be clear how to handle this.

tophmatthews commented:
Maybe @lindsayad can suggest the proper way to use `extra_tag_vectors` and reference residual with mortar method. As for 2DRz, there is hopefully tests already in place for mortar method using axisymmetric that you can reproduce using the action. If not they, should be added

lindsayad commented:
There is an RZ test for thermal mortar contact but not for mechanical

lindsayad commented:
> Maybe @lindsayad can suggest the proper way to use `extra_tag_vectors` and reference residual with mortar method.

I would add a parameter called `extra_vector_tags` with type `TagName`. You can see examples of this for the SolidMechanics and TensorMechanics actions in #14061. You would only want to set the tags for the constraints which are writing into the LM variable residuals though because historically we only save_in from the kernels for the displacement/temperature variables.

shaneongit commented:
Any ideas on the assertion and/or AugmentSparsityOnInterface (maybe same issue)?

lindsayad commented:
You will need to add this code:
```c++
void
ContactAction::addRelationshipManagers(Moose::RelationshipManagerType input_rm_type)
{
  if (are_we_doing_mortar_expr)
    addRelationshipManagers(input_rm_type, validParams<MortarConstraintBase>());
}
```
and then the `AugmentSparsityOnInterface` should get set-up

shaneongit commented:
> You will need to add this code:
> 
> ```c++
> void
> ContactAction::addRelationshipManagers(Moose::RelationshipManagerType input_rm_type)
> {
>   if (are_we_doing_mortar_expr)
>     addRelationshipManagers(input_rm_type, validParams<MortarConstraintBase>());
> }
> ```
> 
> and then the `AugmentSparsityOnInterface` should get set-up

It tuns out the issue was that "component" in NormalMortarMechanicalContact was an enum (x/y/z) while in ContactAction "component" already had a different definition as unsigned int. Since all of the other objects use unsigned for this purpose (including MortarPeriodicAction), I changed NormalMortarMechanicalContact to use the unsigned value. If you think this is a bad idea, let me know.

The addRelationshipManagers call was not necessary. The action now works as expected for normal contact.

shaneongit commented:
@permcody LowerDBlockFromSideset has been marked deprecated, but there are currently no examples of how to use LowerDBlockFromSidesetGenerator in an action (e.g. MortarPeriodicAction). The generator takes a mesh generator name as input, but the file mesh does not seem to be available at the time it is needed by the generator. Any pointers as to how I would replace LowerDBlockFromSideset with LowerDBlockFromSidesetGenerator in this MR?

lindsayad commented:
So you’re telling me that AugmentSparsityOnInterface is getting added when using the action without having the code I recommended? That would surprise me. 

> On Sep 22, 2019, at 3:25 PM, shaneongit <notifications@github.com> wrote:
> 
> You will need to add this code:
> 
> void
> ContactAction::addRelationshipManagers(Moose::RelationshipManagerType input_rm_type)
> {
>   if (are_we_doing_mortar_expr)
>     addRelationshipManagers(input_rm_type, validParams<MortarConstraintBase>());
> }
> and then the AugmentSparsityOnInterface should get set-up
> 
> It tuns out the issue was that "component" in NormalMortarMechanicalContact was an enum (x/y/z) while in ContactAction "component" already had a different definition as unsigned int. Since all of the other objects use unsigned for this purpose (including MortarPeriodicAction), I changed NormalMortarMechanicalContact to use the unsigned value. If you think this is a bad idea, let me know.
> 
> The addRelationshipManagers call was not necessary. The action now works as expected for normal contact.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

shaneongit commented:
> So you’re telling me that AugmentSparsityOnInterface is getting added when using the action without having the code I recommended? That would surprise me.

No, AugmentSparsityOnInterface doesn't appear to be added either way, but the test case works without it (I'm not suggesting it's not needed, but just that it wasn't the source of the error I was seeing). This:
```
addRelationshipManagers(input_rm_type, validParams<MortarConstraintBase>());
```
results in an error: "The requested mortar interface AutomaticMortarGeneration object does not yet exist!"

Changing the params to either _pars or emptyInputParameters() does not result in an error, but AugmentSparsityOnInterface still doesn't show up in the list of objects with --list-constructed-objects.

shaneongit commented:
> > I changed NormalMortarMechanicalContact to use the unsigned value. If you think this is a bad idea, let me know.

> What others have done is setup both a MooseEnum type and a C/C++ enum type. The MooseEnum type is used for obtaining the input from the user. We then initialize the C/C++ enum time for internal use. Ideally this is what you should be doing rather than using a raw unsigned int.

OK, I put it back how it was and just wrote the unsigned to the Enum parameter.

dschwen commented on modules/contact/src/ContactAction.C:
```diff
@@ -97,116 +106,237 @@ ContactAction::ContactAction(const InputParameters & params)
+  if (_formulation == "mortar")
+  {
+    addMortarContact();
+  }
```
curly braces are not required here

dschwen commented on modules/contact/src/ContactAction.C:
```diff
@@ -97,116 +106,237 @@ ContactAction::ContactAction(const InputParameters & params)
+      else if (_system == "DiracKernel")
+      {
+        addDiracContact();
+      }
```
...and here

shaneongit commented:
@lindsayad What is the proper way to check the primal variable order so that the lagrange multiplier order can depend on it? I saw other code checking order of the mesh, but so far haven't been able to query the order "disp_x" even when waiting until after the "add_variable" task.

lindsayad commented:
If you have the `MooseVariable` object, you can do `var.getParamTempl<MooseEnum>("order")`.

shaneongit committed:
Auto-generate new block id for LowerDBlockFromSidesetGenerator.

This enables the ContactAction to generate these side sets
without having the user specify unique IDs for the new
side sets.

shaneongit committed:
Skip mesh generation when recovering.

shaneongit commented:
@permcody @lindsayad To get the mesh generation to work without having the user supply unique IDs for the new side sets, the ID generation must be deferred until side set creation (it can't be easily done inside the action). 

So I modified LowerDBlockFromSidesetGenerator to generate unique IDs if the user does not set `new_block_id`, and there are some obvious -- but hopefully rare -- ways this can fail. Another slightly safer option would be to have something like `auto_generate_new_block_id` that must be set to true if auto generation is expected (otherwise user gets an error). Thoughts?

shaneongit committed:
Set compute_lm_residuals from ContactAction.

lindsayad commented:
I don't think I'm quite clear on the problem. `new_block_id` is a required parameter for `LowerDBlockFromSidesetGenerator`. What do you mean by this:

> if the user does not set `new_block_id`

Can we actually have the `ContactAction` add the `LowerDBlockFromSidesetGenerator` objects? That would be my vote. Mechanics folks are used to having specified sidesets in their incoming `FileMesh(Generator)` or whatever base mesh(generator) object they're using.

shaneongit committed:
Add test for mortar contact action.

shaneongit commented:
Yes, the `ContactAction` now adds the `LowerDBlockFromSidesetGenerator` objects. But to do so, it must generate a unique `SubdomainID` for each sideset, which is not possible because at the time the `MeshGenerator` objects are added, the complete set of existing (user-specified and otherwise) `SubdomainID`s is not known.

So I made `new_block_id` be an optional parameter. If it is not specified, then a unique ID is created when the mesh generator is run and when those new subdomains are instantiated (which is long after `ContactAction` added the `LowerDBlockFromSidesetGenerator` objects).

lindsayad commented:
> at the time the `MeshGenerator` objects are added, the complete set of existing (user-specified and otherwise) `SubdomainID`s is not known.

Ah yes, I've run into this issue before.

> So I made `new_block_id` be an optional parameter. If it is not specified, then a unique ID is created when the mesh generator is run and when those new subdomains are instantiated (which is long after `ContactAction` added the `LowerDBlockFromSidesetGenerator` objects).

This sounds reasonable to me.

> there are some obvious -- but hopefully rare -- ways this can fail

I feel like I'm reading one of those textbooks that says "it can easily be shown". What is is an obvious reason this would fail? If a later mesh generator also wants to use that ID?

shaneongit commented:
> I feel like I'm reading one of those textbooks that says "it can easily be shown". What is is an obvious reason this would fail? If a later mesh generator also wants to use that ID?

Sorry, yes, that's one case. Another is if some other (probably generated) ID is at `std::numeric_limits<SubdomainID>::max()`, in which case the search for another unique ID needs to be a little bit more sophisticated than just adding one to the previous max ID.

lindsayad commented:
Hmm, yea. I think you're doing the best you can do with knowledge you have though. If someone else put in an ID at `std::numeric_limits::max`, then that code is probably the code that needs to be fixed.

shaneongit commented:
> I would add a parameter called `extra_vector_tags` with type `TagName`. You can see examples of this for the SolidMechanics and TensorMechanics actions in #14061. You would only want to set the tags for the constraints which are writing into the LM variable residuals though because historically we only save_in from the kernels for the displacement/temperature variables.

Is this feature possible yet? As discussed in #14061, there is currently no way to set the reference residual for lambda, and there is no way to exclude lambda from the reference residual calculation. 

Anyway, in my testing, an absolute tolerance must be set (`nl_abs_tol`), but so far, this leads to a) bad nonlinear performance and b) some penetration at the interface that appears to be independent of the absolute tolerance. I'm using this test:

https://gist.github.com/shaneongit/3532cfe8c7db934b192be7083e0587e4

lindsayad commented:
@shaneongit Wanted to say that it's pretty great having an external person be able to pick up this stuff and probe so meaningfully into things. Thanks!

> Is this feature possible yet? As discussed in #14061, there is currently no way to set the reference residual for lambda, and there is no way to exclude lambda from the reference residual calculation.

I've opened #14107 which I hope to get to soon. I'll try running your input file and see what I think.

lindsayad commented:
Well I just ran your input file and I thought the non-linear convergence looked great! But I guess a few thoughts:

- You mentioned the penetration...this seems to be an issue with the geometric search. The top node on the LHS is not projecting onto the right-side face. We could probably add some tolerances to the geometric search but that can sometimes be a risky business. What we might better hope for is that the master side is wider than the slave side such that we can be confident that projection will be complete.
- Before contact for this particular problem, `ReferenceResidualProblem` will be ill-suited for a convergence check since at equilibrium the forces everywhere are zero. So here irrespective of whether we are using LM variables to enforce contact we would need an absolute tolerance
- My general thought is that the LM variable should not even be included in the convergence check...if the primal variables are converged (in a PETSc sense), then we are generally happy. Oftentimes when using LM variables, the LM field may even be a bit oscillatory while the primal variable still converges at an optimal rate when considering the L2 or H1 norm (e.g. a finite element sense). So in resolving #14107 I may just want to do no check for the LM variable instead of doing a relative check. But maybe the relative check could still be worth looking into

shaneongit committed:
Add tangential mortar contact to ContactAction.

shaneongit committed:
Added documentation for mortar contact action.

shaneongit commented:
I added tangential, a test, and added to the documentation (in 2 places since the ContactApp stub had nothing at all). 

Everything seems to work fine except for extra_tag_vectors (which should probably wait for #14107) and axisymmetric, which doesn't currently have a test case without the action. Since those two things could easily be added later, I'm removing the WIP status.

shaneongit commented:
This is rebased onto devel as of right now; will the conflict go away when next gets updated?

lindsayad commented:
No, you'll have to rebase onto next and resolve the conflict manually

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/403703/) on e23fb98 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14049/site)

This comment will be updated on new commits.

shaneongit commented:
The failures in (non-mortar) tests are due to 
```registerMooseAction("ContactApp", ContactAction, "add_mesh_generator");```

Apparently, registering the `add_mesh_generator` task turns on checking for the new style of mesh generation, resulting in dozens of these:
```
*** ERROR ***
No generated mesh to retrieve. Your input file should contain either a [Mesh] or [MeshGenerators] 
block.
```
Is adding `type=FileMeshGenerator` to all of these tests the only way to fix this?

lindsayad commented:
What's the backtrace?

tophmatthews commented:
> * My general thought is that the LM variable should not even be included in the convergence check...if the primal variables are converged (in a PETSc sense), then we are generally happy.

That's a pretty bold statement, but it makes sense, and seems key to getting mortar to work on big problems. I have seen good convergence in my primal variables, with the lm with still pretty bad norms
```
30 Nonlinear |R| = 2.289187e-03
    |residual|_2 of individual variables:
              disp_x:        8.87265e-19
              disp_y:        1.68821e-17
              temp:          4.90688e-12
              thermal_lm:    2.70803e-07
              normal_lm:     0.00228919
              tangential_lm: 0
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 30
   Solution, reference convergence variable norms:
   disp_x:        8.87265e-19  disp_x_ref:        1.07889e-08
   disp_y:        1.68821e-17  disp_y_ref:        5.14548e-10
   temp:          4.90688e-12  temp_ref:          38.8304
   thermal_lm:    2.70803e-07  thermal_lm_ref:    0
   normal_lm:     0.00228919  normal_lm_ref:     0
   tangential_lm: 0  tangential_lm_ref: 0
 Solve Did NOT Converge!
```

tophmatthews commented on modules/contact/src/ContactAction.C:
```diff
@@ -97,112 +108,296 @@ ContactAction::ContactAction(const InputParameters & params)
+      _problem->addVariable(var_type, variable_name, params);
+    };
+
+    // Set the family/order same as primal for normal contact and one order lower for tangential.
```
Is this correct? I thought this was supposed to be at least one order lower for both normal and tangential?

shaneongit replied:
Yes, the intended behavior is same order as primal for normal and one lower for tangential. The discussion is up at the top of the thread:
https://github.com/idaholab/moose/pull/14049#issuecomment-532304258

tophmatthews replied:
Gotcha, missed that comment, and I was going off of https://www.mooseframework.org/syntax/Constraints/index.html

lindsayad replied:
See also the responses [here](https://groups.google.com/d/msg/moose-users/JA1yt77MEV0/LX_CgfmwCgAJ)

tophmatthews replied:
The email list was what confused me, since it sounded like you recommend one order lower for thermal, normal, and tangential. Careful re-read is to use an order lower for thermal and tangential, since they are mortar, but same order for node to face discretization.

lindsayad replied:
Yes sir! Although at some point I would really like to do some careful study of the node-face discretization with MMS and explore the finite element convergence rates.

shaneongit commented:
bt:
```
test:glued_contact.test2: *** ERROR ***
test:glued_contact.test2: No generated mesh to retrieve. Your input file should contain either a [Mesh] or [MeshGenerators] block.
test:glued_contact.test2: 
test:glued_contact.test2: 
test:glued_contact.test2: Stack frames: 13
test:glued_contact.test2: 0: libMesh::print_trace(std::ostream&)
test:glued_contact.test2: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
test:glued_contact.test2: 2: void mooseError<char const (&) [105]>(char const (&) [105])
test:glued_contact.test2: 3: MooseApp::getMeshGeneratorMesh(bool)
test:glued_contact.test2: 4: SetupMeshAction::act()
test:glued_contact.test2: 5: Action::timedAct()
test:glued_contact.test2: 6: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
test:glued_contact.test2: 7: ActionWarehouse::executeAllActions()
test:glued_contact.test2: 8: MooseApp::runInputFile()
test:glued_contact.test2: 9: MooseApp::run()
test:glued_contact.test2: 10: /home/shane/projects/bison/moose/modules/combined/combined-opt() [0x402fbe]
test:glued_contact.test2: 11: __libc_start_main
test:glued_contact.test2: 12: /home/shane/projects/bison/moose/modules/combined/combined-opt() [0x40324a]
test:glued_contact.test2: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
test:glued_contact.test2: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
test:glued_contact.test2: :
test:glued_contact.test2: system msg for write_line failure : Bad file descriptor
test:glued_contact.test2: 
test:glued_contact.test2: ################################################################################
test:glued_contact.test2: Tester failed, reason: ERRMSG
test:glued_contact.test2: 
test:glued_contact.test2 ..................................................................... FAILED (ERRMSG)
```

lindsayad commented:
Hmm we might we able to improve the check  in `SetupMeshAction`. Lemme look...

shaneongit commented:
Thanks, Alex. That fixed almost all of the failing tests.

shaneongit commented:
I think this is ready to be merged now, sans the extra_vector_tags and axisymmetric, which can be added under new PRs when those dependencies are met. I'm off contract now, so can't spend as much time on this, but I'll keep an eye on it for code review changes, rebases, etc.

moosebuild commented:
Job [Modules debug PETSc submodule](https://civet.inl.gov/job/404835/) on dfd5c04 : invalidated by @lindsayad

lindsayad commented on modules/contact/test/tests/bouncing-block-contact/gold/frictional-nodal-min-normal-lm-mortar-fb-tangential-lm-mortar-action_out.e:
```diff
@@ -0,0 +1 @@
+frictional-nodal-min-normal-lm-mortar-fb-tangential-lm-mortar-disp_out.e
```
It looks like the non-contact test was re-golded. Was the reason for that the renaming of the variables? Did you verify that the results didn't actually change?

shaneongit replied:
Yes, I checked the results visually and they looked the same. I suppose there's probably a way to exodiff or extract the raw values and compare the renamed vars to the old ones. Is that necessary here?

lindsayad replied:
You can use a custom comparison file like that found [here](https://github.com/idaholab/moose/blob/next/modules/combined/test/tests/sliding_block/in_and_out/constraint/frictionless_kinematic_out.cmp) and you can actually delete the rows corresponding to variables you don't care about. I guess in this case you could delete all but the displacements from the `.cmp` file since those are what we ultimately care about.

However, if they looked the same, then that is probably good enough.

lindsayad replied:
If you ever do want to generate a comparison file, you can execute a command like this:
```
$MOOSE_DIR/framework/contrib/exodiff/exodiff -summary exodus_file.e 2>&1 | tee comparison_file_name.cmp
```

lindsayad commented on modules/contact/test/tests/mechanical-small-problem/gold/frictionless-nodal-lm-mortar-disp-action_out.e:
```diff
@@ -0,0 +1 @@
+frictionless-nodal-lm-mortar-disp_out.e
```
It looks like the non-contact test was re-golded. Was the reason for that the renaming of the variables? Did you verify that the results didn't actually change?

shaneongit replied:
Yes, same as other re-gold: I checked the results visually.

lindsayad commented on modules/contact/test/tests/bouncing-block-contact/frictional-nodal-min-normal-lm-mortar-fb-tangential-lm-mortar-action.i:
```diff
@@ -0,0 +1,148 @@
+  [contact]
+    type = ContactDOFSetSize
+    variable = frictional_normal_lm
+    subdomain = '3'
```
I believe that you deleted this `block_id`, so you should probably change this to `frictional_slave_subdomain`

shaneongit replied:
I'm not sure where this commit is. github can't find it, so maybe I already fixed it?

lindsayad replied:
If you open this file locally, I think you should still be able to see this line...?

shaneongit replied:
Sorry, I was looking in the wrong place. Found and fixed.

lindsayad reviewed (changes requested):
This is really great work. Mostly just the questions about the re-golding.

The `AugmentSparsityOnInterface` relationship manager is not getting added, but this is still working in parallel conceivably because of the ghosting that is being done for the node-face constraints. Getting the relationship manager to actually add will take a little follow-on work which I can do.

shaneongit committed:
Minor edits due to code review.

shaneongit commented on modules/contact/src/ContactAction.C:
```diff
@@ -97,112 +108,296 @@ ContactAction::ContactAction(const InputParameters & params)
+      const int lm_order = primal_type.order.get_order() - codimension;
 
-  if (_current_task == "add_dirac_kernel")
+      if (primal_type.family == MONOMIAL || (primal_type.family == LAGRANGE && lm_order < 1))
```
Good catch. I added a clamp to zero to allow mortar with 0th order elements since a 0th order element should work (shouldn't it?). But if you prefer an error instead, I will fix.

lindsayad replied:
Umm, there are two different concepts from libMesh. The geometric element will have a minimum of first order...e.g. a first order Lagrange polynomial can map between reference and physical coordinates. If an element is second order, you would need a second order Lagrange polynomial to do the mapping, etc.

However, the basis functions associated with approximation solutions can be all sorts of things. Lagrange basis functions have a minimum of first order. Monomial basis functions can be zeroth order, e.g. constant across an element.

lindsayad replied:
Yes this clamp looks fine

shaneongit commented:
Should I be marking these as resolved, or is that up to the reviewer?

lindsayad commented:
> Should I be marking these as resolved, or is that up to the reviewer?

Either is fine. I can always click on "show resolved"

lindsayad reviewed (approved):
This is a great contribution. Thank you @shaneongit
